File: src/SilksongLauncher.Launcher/app/src/main/kotlin/dev/silksong/launcher/skins/core/SkinRotationModels.kt
package dev.silksong.launcher.skins.core

import dev.silksong.launcher.skins.registry.*
import java.util.Collections
import java.text.Normalizer

/** Descriptor-supplied membership/current object, not a catalog or live-visual proof. */
data class RotationDescriptor(val name: String, val currentObject: ActiveVisual.Pack, val eligible: Boolean)
class RotationRing private constructor(entries: List<RotationDescriptor>) {
    val entries: List<RotationDescriptor> = Collections.unmodifiableList(entries.toList())
    companion object {
        fun tryCreate(entries: List<RotationDescriptor>): RotationRing? {
            if (entries.size > 64) return null
            val snapshot = entries.toList()
            if (snapshot.any { !RotationValues.descriptor(it) } ||
                snapshot.map { it.currentObject.id }.distinct().size != snapshot.size) return null
            return RotationRing(snapshot.sortedWith { a, b ->
                val name = RotationValues.utf8Compare(a.name, b.name)
                if (name != 0) name else a.currentObject.id.compareTo(b.currentObject.id)
            })
        }
    }
}
enum class RotationPendingPhase { AWAITING_STABILITY, INTENT_ISSUED }
/** Readiness value only: grants no Prepare/Arm/Apply or publication authority. */
data class RotationReadyIntent(val epoch: DeathEpoch, val hero: HeroBindingToken, val skin: SkinBindingToken,
    val candidate: RotationDescriptor, val prior: ActivationSnapshot)
data class PendingRotation(val candidate: RotationDescriptor, val epoch: DeathEpoch,
    val hero: HeroBindingToken, val skin: SkinBindingToken,
    val phase: RotationPendingPhase = RotationPendingPhase.AWAITING_STABILITY,
    val issuedIntent: RotationReadyIntent? = null)
/** Activation is supplied history. This slice never changes modes or establishes live proof. */
data class RotationState(val ring: RotationRing, val activation: ActivationSnapshot,
    val currentHero: HeroBindingToken? = null, val currentSkin: SkinBindingToken? = null,
    val epochHighWater: DeathEpoch = DeathEpoch(0uL), val pending: PendingRotation? = null)
sealed interface RotationEvent {
    data class Rebind(val hero: HeroBindingToken, val skin: SkinBindingToken) : RotationEvent
    data class ConfirmDeath(val epoch: DeathEpoch, val hero: HeroBindingToken, val skin: SkinBindingToken) : RotationEvent
    data class StableRespawn(val token: StableRespawnToken) : RotationEvent
}
internal object RotationValues {
    private val id = Regex("[a-z0-9](?:[a-z0-9._-]{0,62}[a-z0-9])?")
    private val hash = Regex("[0-9a-f]{64}")
    fun packId(value: String) = value.length <= 64 && id.matches(value)
    fun pack(value: ActiveVisual.Pack) = packId(value.id) &&
        hash.matches(value.treeSha256) && hash.matches(value.contentSha256) && hash.matches(value.importReceiptSha256)
    // Same scalar/name restrictions as CanonicalJson.requireDisplayName and descriptor validation.
    fun text(value: String, maximum: Int): Boolean {
        if (value.isEmpty() || value.length > maximum * 2 || value != value.trim()) return false
        var i = 0; var count = 0
        while (i < value.length) {
            val c = value[i]
            if (c.isISOControl() || c == '؜' || c == '‎' || c == '‏' || c in '‪'..'‮' || c in '⁦'..'⁩') return false
            if (c.isHighSurrogate()) { i++; if (i >= value.length || !value[i].isLowSurrogate()) return false }
            else if (c.isLowSurrogate()) return false
            i++; count++
        }
        return count <= maximum && Normalizer.isNormalized(value, Normalizer.Form.NFKC)
    }
    fun descriptor(value: RotationDescriptor) = text(value.name, 80) && '/' !in value.name && '\\' !in value.name && pack(value.currentObject)
    fun activation(value: ActivationSnapshot) = value.skinStamp >= 0 &&
        (value.selectedPackId == null || packId(value.selectedPackId)) &&
        (value.active is ActiveVisual.Vanilla || value.active is ActiveVisual.Pack && pack(value.active))
    fun utf8Compare(a: String, b: String): Int {
        val left = a.toByteArray(Charsets.UTF_8); val right = b.toByteArray(Charsets.UTF_8)
        for (i in 0 until minOf(left.size, right.size)) {
            val difference = (left[i].toInt() and 255) - (right[i].toInt() and 255)
            if (difference != 0) return difference
        }
        return left.size.compareTo(right.size)
    }
}

/** Fresh trusted caller verification, qualified by both authorities; never derived from history. */
data class HeroVerifiedVisual(val hero: HeroBindingToken, val proof: VerifiedLiveVisualProof)
data class ModeCorrelation(val operationId: String, val hero: HeroBindingToken, val skin: SkinBindingToken)
enum class ModeCommitPhase { ISSUED, INDETERMINATE, DEFINITIVE_FAILURE }
data class ModeOperation(val correlation: ModeCorrelation, val baseHead: VerifiedRegistryHead,
    val target: ActivationSnapshot, val vanillaProof: HeroVerifiedVisual? = null,
    val phase: ModeCommitPhase = ModeCommitPhase.ISSUED, val reboundBlocked: Boolean = false,
    val selectedProof: HeroVerifiedVisual? = null)
/** H1 owns only mode-only CAS. Visual transactions and rebound resolution belong to task66. */
data class RotationModeState(val rotation: RotationState, val head: VerifiedRegistryHead,
    val liveProof: HeroVerifiedVisual? = null, val operationHighWater: Long = 0,
    val operation: ModeOperation? = null, val offRequested: Boolean = false,
    val canceledPending: PendingRotation? = null)
sealed interface RotationModeEvent {
    data object AdvanceMode : RotationModeEvent
    data class VerifiedVisual(val value: HeroVerifiedVisual) : RotationModeEvent
    data class Selector(val event: RotationEvent) : RotationModeEvent
    data class ModeCommitted(val correlation: ModeCorrelation, val receipt: RegistryCommitReceipt,
        val head: VerifiedRegistryHead) : RotationModeEvent
    data class ModeIndeterminate(val correlation: ModeCorrelation) : RotationModeEvent
    data class ModeFailed(val correlation: ModeCorrelation) : RotationModeEvent
    /** Re-emits the identical CAS value, never changes its parent or clears uncertainty. */
    data class RetryModeCommit(val correlation: ModeCorrelation) : RotationModeEvent
}
sealed interface RotationModeCommand {
    val operation: ModeOperation
    data class CommitVerifiedSelectedOn(override val operation: ModeOperation) : RotationModeCommand
    data class CommitOnToRotate(override val operation: ModeOperation) : RotationModeCommand
    data class CommitVerifiedVanillaOff(override val operation: ModeOperation) : RotationModeCommand
}
class RotationModeDecision(val state: RotationModeState, val diagnosis: String,
    vararg commands: RotationModeCommand) {
    val commands: List<RotationModeCommand> = Collections.unmodifiableList(commands.toList())
}

class RotationDecision(val state: RotationState, val diagnosis: String, vararg intents: RotationReadyIntent) {
    val intents: List<RotationReadyIntent> = Collections.unmodifiableList(intents.toList())
}

File: tools/hollow-knight-patches/src/skins/core/SkinRotationModels.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace DualSouls.Skins.HollowKnight.Core
{
    // Descriptor-supplied membership/current object; not catalog or live-visual proof.
    public sealed record RotationDescriptor(string Name, ActiveVisual.Pack CurrentObject, bool Eligible);
    public sealed class RotationRing
    {
        public IReadOnlyList<RotationDescriptor> Entries { get; }
        private RotationRing(IEnumerable<RotationDescriptor> entries) { Entries = Array.AsReadOnly(entries.ToArray()); }
        public static RotationRing TryCreate(IReadOnlyList<RotationDescriptor> entries)
        {
            if (entries == null || entries.Count > 64) return null;
            var snapshot = entries.ToArray();
            if (snapshot.Any(x => !RotationValues.Descriptor(x)) || snapshot.Select(x => x.CurrentObject.Id).Distinct().Count() != snapshot.Length) return null;
            Array.Sort(snapshot, (a,b) => {
                var name = RotationValues.Utf8Compare(a.Name,b.Name);
                return name != 0 ? name : string.CompareOrdinal(a.CurrentObject.Id,b.CurrentObject.Id);
            });
            return new RotationRing(snapshot);
        }
    }
    public enum RotationPendingPhase { AWAITING_STABILITY, INTENT_ISSUED }
    // Readiness only: grants no Prepare/Arm/Apply or publication authority.
    public sealed record RotationReadyIntent(DeathEpoch Epoch, HeroBindingToken Hero, SkinBindingToken Skin,
        RotationDescriptor Candidate, ActivationSnapshot Prior);
    public sealed record PendingRotation(RotationDescriptor Candidate, DeathEpoch Epoch, HeroBindingToken Hero,
        SkinBindingToken Skin, RotationPendingPhase Phase = RotationPendingPhase.AWAITING_STABILITY,
        RotationReadyIntent IssuedIntent = null);
    // Activation is supplied history. This slice never changes modes or establishes live proof.
    public sealed record RotationState(RotationRing Ring, ActivationSnapshot Activation, HeroBindingToken CurrentHero = null,
        SkinBindingToken CurrentSkin = null, DeathEpoch EpochHighWater = null, PendingRotation Pending = null)
    {
        public DeathEpoch EpochHighWater { get; init; } = EpochHighWater ?? new DeathEpoch(0);
    }
    public abstract record RotationEvent
    {
        private RotationEvent() { }
        public sealed record Rebind(HeroBindingToken Hero, SkinBindingToken Skin) : RotationEvent;
        public sealed record ConfirmDeath(DeathEpoch Epoch, HeroBindingToken Hero, SkinBindingToken Skin) : RotationEvent;
        public sealed record StableRespawn(StableRespawnToken Token) : RotationEvent;
    }
    internal static class RotationValues
    {
        private static readonly Regex Id = new("\\A[a-z0-9](?:[a-z0-9._-]{0,62}[a-z0-9])?\\z");
        private static readonly Regex Hash = new("\\A[0-9a-f]{64}\\z");
        internal static bool PackId(string value) => value != null && value.Length <= 64 && Id.IsMatch(value);
        internal static bool Digest(string value) => value != null && value.Length == 64 && Hash.IsMatch(value);
        internal static bool Pack(ActiveVisual.Pack value) => value != null && PackId(value.Id) &&
            Digest(value.TreeSha256) && Digest(value.ContentSha256) && Digest(value.ImportReceiptSha256);
        // Same scalar/name restrictions as Kotlin CanonicalJson.requireDisplayName.
        internal static bool Text(string value, int maximum)
        {
            if (string.IsNullOrEmpty(value) || value.Length > maximum * 2 || value != value.Trim()) return false;
            var count = 0;
            for (var i=0; i<value.Length; i++,count++)
            {
                var c=value[i];
                if (char.IsControl(c) || c=='؜' || c=='‎' || c=='‏' || c>='‪' && c<='‮' || c>='⁦' && c<='⁩') return false;
                if (char.IsHighSurrogate(c)) { i++; if (i>=value.Length || !char.IsLowSurrogate(value[i])) return false; }
                else if (char.IsLowSurrogate(c)) return false;
            }
            return count <= maximum && NormalizedScalarText(value);
        }
        // Called only after the bounded full-string surrogate/control validation above.
        // CLR normalization rejects some valid noncharacters (notably U+FFFE), unlike
        // Java NFKC. All 66 noncharacters have no decomposition and CCC=0, so they
        // separate normalization runs without changing the original text or UTF-8 order.
        private static bool NormalizedScalarText(string value)
        {
            var start = 0;
            for (var i = 0; i < value.Length;)
            {
                var scalar = char.ConvertToUtf32(value, i);
                var width = scalar > 0xFFFF ? 2 : 1;
                var noncharacter = (scalar >= 0xFDD0 && scalar <= 0xFDEF) || (scalar & 0xFFFF) >= 0xFFFE;
                if (noncharacter)
                {
                    if (!NormalizedSegment(value, start, i - start)) return false;
                    start = i + width;
                }
                i += width; // A supplementary boundary consumes the complete surrogate pair.
            }
            return NormalizedSegment(value, start, value.Length - start);
        }
        private static bool NormalizedSegment(string value, int start, int length)
        {
            if (length == 0) return true;
            try
            {
                return value.Substring(start, length).IsNormalized(NormalizationForm.FormKC);
            }
            catch (ArgumentException)
            {
                // An unexpected ordinary-segment rejection never becomes normalization proof.
                return false;
            }
        }
        internal static bool Descriptor(RotationDescriptor value) => value != null && Text(value.Name,80) &&
            value.Name.IndexOf('/') < 0 && value.Name.IndexOf('\\') < 0 && Pack(value.CurrentObject);
        internal static bool Activation(ActivationSnapshot value) => value != null &&
            (value.Mode==SkinMode.OFF || value.Mode==SkinMode.ON || value.Mode==SkinMode.ROTATE) && value.SkinStamp>=0 &&
            (value.SelectedPackId==null || PackId(value.SelectedPackId)) &&
            (value.Active is ActiveVisual.Vanilla || value.Active is ActiveVisual.Pack p && Pack(p));
        internal static int Utf8Compare(string a,string b)
        {
            var left=Encoding.UTF8.GetBytes(a);var right=Encoding.UTF8.GetBytes(b);
            for(var i=0;i<Math.Min(left.Length,right.Length);i++) { var difference=left[i]-right[i];if(difference!=0)return difference; }
            return left.Length.CompareTo(right.Length);
        }
    }
    // Value mirror of registry proof. Freshness is a trusted event boundary, not persisted history.
    public abstract record VerifiedLiveVisualProof(SkinBindingToken Binding)
    {
        public sealed record Vanilla(SkinBindingToken Binding) : VerifiedLiveVisualProof(Binding);
        public sealed record Pack(SkinBindingToken Binding, ActiveVisual.Pack Visual) : VerifiedLiveVisualProof(Binding);
    }
    public sealed record HeroVerifiedVisual(HeroBindingToken Hero, VerifiedLiveVisualProof Proof);
    public sealed record ModeCorrelation(string OperationId, HeroBindingToken Hero, SkinBindingToken Skin);
    public enum ModeCommitPhase { ISSUED, INDETERMINATE, DEFINITIVE_FAILURE }
    public sealed record ModeOperation(ModeCorrelation Correlation, VerifiedRegistryHead BaseHead,
        ActivationSnapshot Target, HeroVerifiedVisual VanillaProof = null,
        ModeCommitPhase Phase = ModeCommitPhase.ISSUED, bool ReboundBlocked = false, HeroVerifiedVisual SelectedProof = null);
    // H1 owns only mode-only CAS. Visual transactions and rebound resolution belong to task66.
    public sealed record RotationModeState(RotationState Rotation, VerifiedRegistryHead Head,
        HeroVerifiedVisual LiveProof = null, long OperationHighWater = 0, ModeOperation Operation = null,
        bool OffRequested = false, PendingRotation CanceledPending = null);
    public abstract record RotationModeEvent
    {
        private RotationModeEvent() { }
        public sealed record AdvanceMode : RotationModeEvent;
        public sealed record VerifiedVisual(HeroVerifiedVisual Value) : RotationModeEvent;
        public sealed record Selector(RotationEvent Event) : RotationModeEvent;
        public sealed record ModeCommitted(ModeCorrelation Correlation, RegistryCommitReceipt Receipt, VerifiedRegistryHead Head) : RotationModeEvent;
        public sealed record ModeIndeterminate(ModeCorrelation Correlation) : RotationModeEvent;
        public sealed record ModeFailed(ModeCorrelation Correlation) : RotationModeEvent;
        // Retry is the identical CAS value, never a changed parent or reset of uncertainty.
        public sealed record RetryModeCommit(ModeCorrelation Correlation) : RotationModeEvent;
    }
    public abstract record RotationModeCommand(ModeOperation Operation)
    {
        public sealed record CommitVerifiedSelectedOn(ModeOperation Operation) : RotationModeCommand(Operation);
        public sealed record CommitOnToRotate(ModeOperation Operation) : RotationModeCommand(Operation);
        public sealed record CommitVerifiedVanillaOff(ModeOperation Operation) : RotationModeCommand(Operation);
    }
    public sealed class RotationModeDecision
    {
        public RotationModeState State { get; }
        public string Diagnosis { get; }
        public IReadOnlyList<RotationModeCommand> Commands { get; }
        public RotationModeDecision(RotationModeState state, string diagnosis, params RotationModeCommand[] commands)
        { State = state; Diagnosis = diagnosis; Commands = Array.AsReadOnly((RotationModeCommand[])commands.Clone()); }
    }
    public sealed class RotationDecision
    {
        public RotationState State { get; }
        public string Diagnosis { get; }
        public IReadOnlyList<RotationReadyIntent> Intents { get; }
        public RotationDecision(RotationState state, string diagnosis, params RotationReadyIntent[] intents)
        { State = state; Diagnosis = diagnosis; Intents = Array.AsReadOnly((RotationReadyIntent[])intents.Clone()); }
    }
}
